src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -1309,7 +1309,12 @@ public function registerForAutoconfiguration(string $interface)
     /**
      * Registers an attribute that will be used for autoconfiguring annotated classes.
      *
-     * The configurator will receive a ChildDefinition instance, an instance of the attribute and the corresponding \ReflectionClass, in that order.
+     * If an attribute can only be used on 1 target, the \Reflector type should be narrowed to
+     * one of: \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter
+     *
+     * If an attribute can be used on multiple targets, a union type can be used based on above types.
+     *
+     * @param callable(ChildDefinition $definition, object $attribute, \Reflector $reflector): void $configurator
```

Comment:
Psalm will also complain with this, with a userland configurator will use eg `ReflectionMethod $r`...

Reply:
then they can ignore the error or they \assert($reflector instanceof ReflectionMethod) inside the callable instead. To me autocomplete and proper signature documentation is the right thing to do.

Also this doc should be
```
@template T
@param class-string<T> $attributeClass
@param callable(ChildDefinition $definition, T $attribute, \Reflector $reflector): void $configurator
```

Reply:
As @ruudk mentioned earlier, the type-hint is used as a filter to skip scanning attributes on eg parameters when no attribute is expected on them. It's true that we could improve the filter and also consider the attributes on the attribute classes, but I'm not sure we need the extra code, because as I wrote in https://github.com/symfony/symfony/pull/42039#discussion_r670431074, relying on attribute's flags is less flexible than relying on the argument's type hint: it creates an unnecessary coupling between the configurators and the attributes their deal with.

Anyway, I'm fine with the proposed phpdoc, provided we throw an explanatory exception when Reflector is kept as a type-hint.

Reply:
Added the proposed phpdoc and added a guard to prevent usage of `Reflector` argument.